### PR TITLE
Enable or disable localhost client via allowed_clients

### PR DESCRIPTION
**Pull request by chatton**
## Description

closes: #2537

The issue referenced mentions adding a new `enabled` field for localhost client, however this didn't fit super cleanly with the existing fields as we already have a list of allowed clients. This PR instead uses the presence/absence of the localhost client ID in this array to return the `Status` of the client as `Active`/`Frozen` respectively. 

### Commit Message / Changelog Entry

N/A (included in feature release notes)

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/02-client/keeper/keeper.go [hunk 1]
```diff
@@ -31,10 +34,11 @@ type Keeper struct {
 	paramSpace    paramtypes.Subspace
 	stakingKeeper types.StakingKeeper
 	upgradeKeeper types.UpgradeKeeper
+	authority     string
 }
 
 // NewKeeper creates a new NewKeeper instance
-func NewKeeper(cdc codec.BinaryCodec, key storetypes.StoreKey, paramSpace paramtypes.Subspace, sk types.StakingKeeper, uk types.UpgradeKeeper) Keeper {
+func NewKeeper(cdc codec.BinaryCodec, key storetypes.StoreKey, paramSpace paramtypes.Subspace, sk types.StakingKeeper, uk types.UpgradeKeeper, authority string) Keeper {
```
Comment (damiannolan) [inline modules/core/02-client/keeper/keeper.go:41]: If we're going to go ahead with this approach we should cherry pick this API breaking change out to a separate PR that we can backport to v7

Comment (chatton) [reply modules/core/02-client/keeper/keeper.go:41]: do you just mean the method signature change? Not the `MsgUpdateParams` ?

Comment (damiannolan) [reply modules/core/02-client/keeper/keeper.go:41]: Yeah, just the method signature change and addition of the `authority` field on the keeper struct

File: modules/core/02-client/keeper/keeper_test.go [hunk 2]
```diff
@@ -461,3 +463,73 @@ func (suite KeeperTestSuite) TestIterateClientStates() { //nolint:govet // this
...
+			msgUpdateParams := &types.MsgUpdateParams{}
+			tc.malleate(msgUpdateParams)
+
+			_, err := ibcKeeper.ClientKeeper.UpdateParams(ctxA, msgUpdateParams)
+
+			clientParams = ibcKeeper.ClientKeeper.GetParams(ctxA)
+			cs = suite.chainA.GetClientState(exported.Localhost)
+			localhostClientState = cs.(*localhost.ClientState)
+
+			suite.Require().Equal(clientParams.IsAllowedClient(exported.Localhost), localhostClientState.Enabled)
```
Comment (crodriguezvega) [inline modules/core/02-client/keeper/keeper_test.go:526]: Wouldn't it be more safe to check both `clientParams.IsAllowedClient(exported.Localhost)` and `localhostClientState.Enabled` against an expected value? Otherwise we could have the situation were `UpdateParams` doesn't do what it's supposed to do when disabling the params (neither params get updated nor client state gets updated) and the test would still pass.

File: proto/ibc/lightclients/localhost/v2/localhost.proto [hunk 3]
```diff
@@ -13,4 +13,7 @@ message ClientState {
 
   // the latest block height
   ibc.core.client.v1.Height latest_height = 1 [(gogoproto.nullable) = false];
+
+  // whether or not the localhost client should be enabled
```
Comment (crodriguezvega) [inline proto/ibc/lightclients/localhost/v2/localhost.proto:17]:
```suggestion
  // whether or not the localhost client is enabled
```

File: modules/core/02-client/keeper/keeper.go [hunk 4]
```diff
@@ -403,3 +408,25 @@ func (k Keeper) ClientStore(ctx sdk.Context, clientID string) sdk.KVStore {
...
 }
+
+// UpdateParams implements MsgServer.UpdateParams method.
+// It defines a method to update the 02-client module parameters.
+func (k Keeper) UpdateParams(goCtx context.Context, req *types.MsgUpdateParams) (*types.MsgUpdateParamsResponse, error) {
+	if k.authority != req.Authority {
+		return nil, errors.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", k.authority, req.Authority)
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	k.SetParams(ctx, req.Params)
```
Comment (crodriguezvega) [inline modules/core/02-client/keeper/keeper.go:419]: This is a nit, but should we move this call to the end once all the checks have passed?

File: modules/light-clients/09-localhost/client_state.go [hunk 5]
```diff
@@ -57,11 +61,7 @@ func (cs ClientState) Initialize(ctx sdk.Context, cdc codec.BinaryCodec, clientS
 		return sdkerrors.Wrap(clienttypes.ErrInvalidConsensus, "initial consensus state for localhost must be nil.")
 	}
 
-	clientState := ClientState{
```
Comment (crodriguezvega) [inline modules/light-clients/09-localhost/client_state.go:60]: Why was this removed?

**Review by crodriguezvega — COMMENTED** [review]
Nice work, @chatton! Some of this work can also partially count for https://github.com/cosmos/ibc-go/issues/2010 :)

I think we should add in this PR or in the one @damiannolan is working on documentation for enabling/disabling the localhost.

File: modules/light-clients/09-localhost/client_state.go (hunk 5 above)
Comment (chatton) [reply modules/light-clients/09-localhost/client_state.go:60]: Because `Initialize` is an interface function, and previously we were not actually using the method receiver. The usage looks like

```go
func (k Keeper) CreateLocalhostClient(ctx sdk.Context) error {
	clientState := localhost.NewClientState(types.GetSelfHeight(ctx), k.GetParams(ctx).IsAllowedClient(exported.Localhost))
	return clientState.Initialize(ctx, k.cdc, k.ClientStore(ctx, exported.Localhost), nil)
}
```

This enables us to configure the clientState.Enabled field based on the `allowed_clients` list.

File: modules/core/02-client/keeper/keeper_test.go (hunk 2 above)
Comment (chatton) [reply modules/core/02-client/keeper/keeper_test.go:526]: yes I think that's a good idea!

File: modules/core/02-client/keeper/keeper.go (hunk 4 above)
Comment (chatton) [reply modules/core/02-client/keeper/keeper.go:419]: Sure I think that makes sense!

**Comment by damiannolan** [comment]
I think it would be good to update the localhost client update logic [here](https://github.com/cosmos/ibc-go/blob/09-localhost/modules/core/02-client/keeper/keeper.go#L64-L66) to check the status. This would prevent unnecessary writes to state when the localhost client is disabled.

```go
func (k Keeper) UpdateLocalhostClient(ctx sdk.Context, clientState exported.ClientState) []exported.Height {
	if status := clientState.Status(ctx, k.ClientStore(ctx, exported.Localhost), k.cdc); status == exported.Active {
		clientState.UpdateState(ctx, k.cdc, k.ClientStore(ctx, exported.Localhost), nil)
	}

	return nil
}
```

**Comment by chatton** [comment]
closing in favour of #3196
